pkg/kubelet/status/manager.go

```diff
@@ -392,6 +405,94 @@ func (m *manager) needsUpdate(uid types.UID, status versionedPodStatus) bool {
 	return !ok || latest < status.version
 }
 
+// needsReconcile compares the cached status with the status in the pod manager (which
+// in fact comes from apiserver), returns whether the status needs to be reconciled with
+// the apiserver. Now when pod status is inconsistent between apiserver and kubelet,
+// kubelet should forcibly send an update to reconclie the inconsistence, because kubelet
+// should be the source of truth of pod status.
+func (m *manager) needsReconcile(uid types.UID, status api.PodStatus) bool {
+	// The pod could be a static pod, so we should translate first.
+	pod, ok := m.podManager.GetPodByUID(uid)
+	if !ok {
+		// Although we get uid from pod manager in syncBatch, it still could be deleted before here.
+		glog.V(4).Infof("Pod %q has been deleted, no need to reconcile", format.Pod(pod))
+		return false
+	}
+	// If the pod is a static pod, we should check its mirror pod, because only status in mirror pod is meaningful to us.
+	if kubepod.IsStaticPod(pod) {
+		mirrorPod, ok := m.podManager.GetMirrorPodByPod(pod)
+		if !ok {
+			glog.V(4).Infof("Static pod %q has no corresponding mirror pod, no need to reconcile", format.Pod(pod))
+			return false
+		}
+		pod = mirrorPod
+	}
+
+	// Normalize the status from pod manager, in fact the status should have been normalized (by apiserver), that's
+	// why we have to normalize the cached status before comparing here. However, in case that the implementation on
+	// apiserver may be changed someday, also normalize the status passed in here.
+	// TODO: (random-liu) When timestamp is made consistent on client and server side, remove all the following
+	// normalization.
+	normalizedStatus := normalizeStatus(pod.Status)
+	// Normalize the cached status
+	normalizedCachedStatus := normalizeStatus(status)
+	if isStatusEqual(&normalizedStatus, &normalizedCachedStatus) {
+		// If the status from the source is the same with the cached status,
+		// reconcile is not needed. Just return.
+		return false
+	}
+	glog.V(3).Infof("Pod status is inconsistent with cached status, a reconcile should be triggered: %+v", util.ObjectDiff(normalizedStatus, normalizedCachedStatus))
+
+	return true
+}
+
+// We add this function, because apiserver only supports *RFC3339* now, which means that the timestamp returned by
+// apiserver has no nanosecond infromation. However, the timestamp returned by unversioned.Now() contains nanosecond,
+// so when we do comparation between status from apiserver and cached status, isStatusEqual() will always return false.
+// There is related issue #15262 and PR #15263 about this.
+// In fact, the best way to solve this is to do it on api side. However for now, we normalize the status locally in
+// kubelet temporarily.
+// TODO: (random-liu) Remove this after apiserver makes it consistent on api side.
+func normalizeStatus(cachedStatus api.PodStatus) api.PodStatus {
```

Comment:
Some test such as [this](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/manager_test.go#L212) needs sub-second precision, but we can of course modify the test to let it pass. Since there is no much difference, I think maybe for now we can keep the precision in the cache.